### PR TITLE
Packaging with `pyproject.toml`

### LINKED ISSUE #7: [Proposal] Migrate from `setup.py` to `pyproject.toml`
### Proposal 

I'd like to propose to migrate from using `setup.py` to using `pyproject.toml`.

### Motivation

Probably not super urgent, but since `pyproject.toml` should be the new standard and `setup.py` is now considered legacy ([see `pip` documentation](https://pip.pypa.io/en/stable/reference/build-system/setup-py/#setup-py-legacy)), it could be good to migrate at some point.

### Pitch

- Migrate all the content of `setup.py` to `pyproject.toml`
- Have a single source of truth for all the dependencies in `pyproject.toml`, thus remove `requirements.txt` and `test_requirements.txt`

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/openai/gym/issues) in the repo (**required**)

### LINKED ISSUE #8: [Proposal] Add python 3.11 support
Python 3.11 is meant to be released in Oct 2022 that provides a number of speedup, up to 30%. 
As software is fairly low in the programming stack then we should look to support it soon

### PR DESCRIPTION
**Pull request by XuehaiPan**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Resolves #7
Resolves #8
Closes #86

Repackage project with `pyproject.toml` using `setuptools` backend.

Different from PR #86 using `poetry` as the backend, this PR does not require the users have `poetry` installed. The build dependencies `pip` and `setuptools` are out-of-box available for a Python environment.

It also resolves the `use-pep517` issue in https://github.com/Farama-Foundation/Gymnasium/pull/86#issuecomment-1294321247:

```console
$ python3 -m pip uninstall box2d-py
$ python3 -m pip install -e '.[box2d]'
Looking in indexes: https://pypi.tuna.tsinghua.edu.cn/simple
Obtaining file:///Users/PanXuehai/PycharmProjects/Gymnasium
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Installing backend dependencies ... done
  Preparing editable metadata (pyproject.toml) ... done
Requirement already satisfied: jax-jumpy>=0.2.0 in ./venv/lib/python3.10/site-packages (from gymnasium==0.27.0) (0.2.0)
Requirement already satisfied: shimmy<1.0,>=0.1.0 in ./venv/lib/python3.10/site-packages (from gymnasium==0.27.0) (0.1.0)
Requirement already satisfied: cloudpickle>=1.2.0 in ./venv/lib/python3.10/site-packages (from gymnasium==0.27.0) (2.2.0)
Requirement already satisfied: numpy>=1.21.0 in ./venv/lib/python3.10/site-packages (from gymnasium==0.27.0) (1.23.5)
Requirement already satisfied: gymnasium-notices>=0.0.1 in ./venv/lib/python3.10/site-packages (from gymnasium==0.27.0) (0.0.1)
Requirement already satisfied: pygame==2.1.0 in ./venv/lib/python3.10/site-packages (from gymnasium==0.27.0) (2.1.0)
Requirement already satisfied: swig==4.* in ./venv/lib/python3.10/site-packages (from gymnasium==0.27.0) (4.1.0)
Collecting box2d-py==2.3.5
  Using cached https://pypi.tuna.tsinghua.edu.cn/packages/dd/5a/ad8d3ef9c13d5afcc1e44a77f11792ee717f6727b3320bddbc607e935e2a/box2d-py-2.3.5.tar.gz (374 kB)
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: gymnasium
  Building editable for gymnasium (pyproject.toml) ... done
  Created wheel for gymnasium: filename=gymnasium-0.27.0-0.editable-py3-none-any.whl size=6144 sha256=76aa303939a79f12079492dc95bfe8f158bbd31e62324ea8f9331cea6b1bf8bf
  Stored in directory: /private/var/folders/2k/0fjtszys5v9f8znv9rwzn53m0000gn/T/pip-ephem-wheel-cache-hduqlscc/wheels/42/43/7c/5a225e499cce07ffffb46128b394cad4d1438f70a29fdd29d2
Successfully built gymnasium
Installing collected packages: box2d-py, gymnasium
  DEPRECATION: box2d-py is being installed using the legacy 'setup.py install' method, because it does not have a 'pyproject.toml' and the 'wheel' package is not installed. pip 23.1 will enforce this behaviour change. A possible replacement is to enable the '--use-pep517' option. Discussion can be found at https://github.com/pypa/pip/issues/8559
  Running setup.py install for box2d-py ... done
  Attempting uninstall: gymnasium
    Found existing installation: gymnasium 0.27.0
    Uninstalling gymnasium-0.27.0:
      Successfully uninstalled gymnasium-0.27.0
Successfully installed box2d-py-2.3.5 gymnasium-0.27.0
```

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

# Checklist:

- [X] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,112 @@
+# Package ######################################################################
+
+[build-system]
+requires = ["setuptools >= 61.0.0"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "gymnasium"
```
Comment (XuehaiPan) [inline pyproject.toml:8]: The PyPI project name is lower-cased to fit the [PEP 503: Normalized Names](https://peps.python.org/pep-0503/#normalized-names) convention.

```python
import re

def normalize(name):
    return re.sub(r"[-_.]+", "-", name).lower()
```

File: pyproject.toml [hunk 2]
```diff
@@ -1,3 +1,112 @@
...
+[project.optional-dependencies]
+# Update dependencies in `all` if any are added or removed
+atari = ["shimmy[atari] >=0.1.0,<1.0"]
+accept-rom-license = ["autorom[accept-rom-license] ~=0.4.2"]
+box2d = ["box2d-py ==2.3.5", "pygame ==2.1.0", "swig ==4.*"]
+classic-control = ["pygame ==2.1.0"]
+classic_control = ["pygame ==2.1.0"]  # kept for backward compatibility
+mujoco-py = ["mujoco-py >=2.1,<2.2"]
+mujoco_py = ["mujoco-py >=2.1,<2.2"]  # kept for backward compatibility
+mujoco = ["mujoco >=2.3.0", "imageio >=2.14.1"]
```
Comment (XuehaiPan) [inline pyproject.toml:46]: The extra-require name is also lower-cased to fit the [PEP 503: Normalized Names](https://peps.python.org/pep-0503/#normalized-names) convention as commented above. The old names are kept for backward compatibility:

```bash
pip3 install gymnasium[classic-control]
pip3 install gymnasium[classic_control]
```

File: pyproject.toml [hunk 3]
```diff
@@ -1,3 +1,113 @@
...
+    "tensorflow >=2.1.0",
+    "torch >=1.0.0",
+]
+testing = ["pytest ==7.1.3"]
+
+[project.urls]
+Homepage = "https://farama.org"
+Repository = "https://github.com/Farama-Foundation/Gymnasium"
+Documentation = "https://gymnasium.farama.org"
+"Bug Report" = "https://github.com/Farama-Foundation/Gymnasium/issues"
```
Comment (XuehaiPan) [inline pyproject.toml:94]: Should the `Homepage` be `"https://gymnasium.farama.org"`, the same as `Documentation`?

File: pyproject.toml [hunk 4]
```diff
@@ -1,3 +1,113 @@
...
+    "gymnasium-notices >=0.0.1",
+    "shimmy >=0.1.0,<1.0",
+]
+dynamic = ["version"]
+
+[project.optional-dependencies]
+# Update dependencies in `all` if any are added or removed
+atari = ["shimmy[atari] >=0.1.0,<1.0"]
+accept-rom-license = ["autorom[accept-rom-license] ~=0.4.2"]
+box2d = ["box2d-py ==2.3.5", "pygame ==2.1.0", "swig ==4.*"]
```
Comment (pseudo-rnd-thoughts) [inline pyproject.toml:42]: We have updated pygame to `2.1.2.dev8` I believe which adds support for python 3.11

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Could you update the pygame version for python 3.11 support then good to merge

**Review by pseudo-rnd-thoughts — APPROVED** [review]

**Pull request merged**